### ISSUE TITLE
Direct descendant selector support

### ISSUE THREAD

**Issue by meyer** [jsxstyle/jsxstyle#73]
Sometimes hovering on a thing or clicking on a thing should affect a child element. Creating a separate stylesheet just for this functionality seems a bit silly to me.

I think there’s a nice way we could expose very limited support for descendant selectors.

## Proposal

It’s a two part thing.

### Part 1: the parent

If…
1. a style prop on a `jsxstyle` element is prefixed with a pseudoclass, and…
2. said prop is also prefixed with something indicating that it’s styling a child element (`Child` in this case)…

then add an extra class that has the descendant style plus a unique descendant-y class name.

Code that looks like this:

```jsx
<Block
  component="a"
  hoverChildColor="blue"
  props={{ href: 'https://google.com' }}>
  ...
</Block>
```

becomes this JS:

```jsx
<a
  href="https://google.com"
  className="_jsxstyleClassName1">
   ...
</a>
```

and this CSS:

```css
._jsxstyleClassName1 > ._childClassName {
  color: blue;
}
```

### Part 2: the child

If a component has a prop that indicates it should be affected by descendant styles, then it receives the child className.

This code:

```jsx
<Block
  component={SVGAsset}
  props={{ assetName: 'icon/disclosure-arrow-16' }}
  isChild
/>
```

becomes this JS:

```jsx
<SVGAsset
  assetName="icon/disclosure-arrow-16"
  className="_jsxstyleClassName2 _childClassName"
/>
```

## Why this sucks

1. Only supports direct descendants. I haven’t thought of an ✨ elegant ✨ way to get around this yet but I think it is possible to support descendants. The user would have to specify a descendant key and keep track of it, though, and there’s the possibility of collisions if you’ve got more than one person writing component code. We could add utilities to generate a Symbol-like key though I guess. Hmmmmmm.
2. people will write their code a bit differently so that the thing they want to style is a direct child of the thing they’re interacting with.

Basically, it’s sucks because it’s extremely limited, but that’s also why it’s g[oo][]d IMO.

🤔💡 Thought: we could also add adjacent sibling support this way, too.

Feedback welcome/appreciated.

[oo]: https://twitter.com/dril/status/473265809079693312

**Comment by meyer** [comment]
Thoughts for descendant selector support:

Prop names in these examples are obviously not final.

#### Option A: specify a className

```jsx
const k = jsxstyle.functionThatGetsAUniqueClassName();

<Block hoverChildColor="blue" childClassName={k}>
  <Block>
    <Block>
      <Block className={k}>
        ¯\_(ツ)_/¯
      </Block>
    </Block>
  </Block>
</Block>
```
👍 uses an existing prop
👍 concept makes sense; less mysterious
👎 className can be a known thing, and that could lead to collisions

#### Option B: share a key:

```jsx
const k = jsxstyle.functionThatGetsAUniqueKey();

<Block hoverChildColor="blue" parentKey={k}>
  <Block>
    <Block>
      <Block childKey={k}>
        ¯\_(ツ)_/¯
      </Block>
    </Block>
  </Block>
</Block>
```

👍 final class name isn’t known. The key can be an object or weird string that the component validates.
👎 Two extra props instead of one.

**Comment by meyer** [comment]
Oh yeah, and with either of these, omitting the prop on the parent would make `jsxstyle` fall back to using the direct descendant selector.

**Comment by meyer** [comment]
Closing for now.

**Comment by steadicat** [comment]
Would love this feature. A few more options below.

The way `glamor-jsxstyle` does it, except with unique keys instead of class names:
```js
const k = jsxstyle.functionThatGetsAUniqueKey();

<Block hover={{[k]: {color: 'blue'}}}>
  <Block>
    <Block>
      <Block childKey={k}>
        ¯\_(ツ)_/¯
      </Block>
    </Block>
  </Block>
</Block>
```

Similar to the above, but inverted so the rules are defined closer to where they are applied. The `hover` object becomes a little harder to parse though:
```js
const k = jsxstyle.functionThatGetsAUniqueKey();

<Block parentKey={k}>
  <Block>
    <Block>
      <Block hover={{[k]: {color: 'blue'}}}>
        ¯\_(ツ)_/¯
      </Block>
    </Block>
  </Block>
</Block>
```

This last one is not as powerful (can't have multiple parents targeting a single child), but it's perhaps the cleanest API. One could argue that more complex stuff should be handled with JS state.

```js
<Block isParent>
  <Block>
    <Block>
      <Block parentHoverColor="blue">
        ¯\_(ツ)_/¯
      </Block>
    </Block>
  </Block>
</Block>
```

**Comment by meyer** [comment]
@steadicat Yeah, the simplest API is definitely the best IMO. Right now I can explain the concept of `jsxstyle` in a pretty short sentence and I can explain the particulars in another short sentence. I’d like to keep it that way if at all possible.

I like that parent styles that affect the child are defined on the child in your examples. Makes a lot of sense. I wonder if there’s a way to implement this and keep `jsxstyle` components unaware of each other?

**Comment by meyer** [comment]
I’m not opposed to this idea but I’m closing this issue as I haven’t really needed this feature in any of my projects.
